core/cog-directory-files-handler.c

```diff
@@ -145,26 +147,45 @@ cog_directory_files_handler_run (CogRequestHandler      *request_handler,
 {
     CogDirectoryFilesHandler *handler = COG_DIRECTORY_FILES_HANDLER (request_handler);
 
+    g_autoptr(SoupURI) uri =
+        soup_uri_new (webkit_uri_scheme_request_get_uri (request));
+
     /*
      * If we get an empty path, redirect to the root resource "/", otherwise
      * subresources cannot load properly as there would be no base URI.
      */
-    const char *path = webkit_uri_scheme_request_get_path (request);
+    const char *path = soup_uri_get_path (uri);
     if (path[0] != '/') {
-        g_autofree char *uri =
-            g_strconcat (webkit_uri_scheme_request_get_scheme (request), ":/", path, NULL);
-        webkit_web_view_load_uri (webkit_uri_scheme_request_get_web_view (request), uri);
+        soup_uri_set_path (uri, "/");
+        g_autofree char *uri_string = soup_uri_to_string (uri, FALSE);
+        webkit_web_view_load_uri (webkit_uri_scheme_request_get_web_view (request), uri_string);
         return;
     }
 
+    g_autoptr(GFile) base_path = NULL;
+    if (handler->use_host) {
+        const char *host = soup_uri_get_host (uri);
+        if (!host || host[0] == '\0') {
+            g_autofree char *uri_string = soup_uri_to_string (uri, FALSE);
+            g_autoptr(GError) error = g_error_new (G_FILE_ERROR,
+                                                   G_FILE_ERROR_INVAL,
+                                                   "No host in URI: %s",
+                                                   uri_string);
+            webkit_uri_scheme_request_finish_error (request, error);
+            return;
+        }
+        base_path = g_file_get_child (handler->base_path, host);
+    }
```

Comment:
I would probably just add an else here and assign `base_path` the handler's base path, to avoid having the ?: comparison later on, keeping the refcounting in mind of course.

Reply:
Fun thing: First I wrote it in the way you suggest, then I thought that I would change it to use `?:` below and avoid refcount churn :smile: 

I do agree that the code is easier to understand the way you suggest, so I'll change it back to avoid the `?:` before merging.

Thanks for the review!